dex/networks/eth/contractor.go

```diff
@@ -0,0 +1,192 @@
+	contractAddr, exists := ContractAddresses[0][net]
+	if !exists || contractAddr == (common.Address{}) {
+		return nil, fmt.Errorf("no contract address for version 0, net %s", net)
+	}
+	c, err := NewETHSwap(contractAddr, ec)
+	if err != nil {
+		return nil, err
+	}
+	parsedABI, err := abi.JSON(strings.NewReader(ETHSwapABI))
+	if err != nil {
+		return nil, err
+	}
+	return &contractorV0{
+		ETHSwap:      c,
+		abi:          parsedABI,
+		ec:           ec,
+		contractAddr: contractAddr,
+		acctAddr:     acctAddr,
+	}, nil
+}
+
+func (c *contractorV0) Initiate(txOpts *bind.TransactOpts, contracts []*asset.Contract) (*types.Transaction, error) {
+	inits := make([]ETHSwapInitiation, 0, len(contracts))
+	for _, contract := range contracts {
+		var secretHash [32]byte
+		copy(secretHash[:], contract.SecretHash)
+		bigVal := big.NewInt(0).SetUint64(contract.Value)
+		inits = append(inits, ETHSwapInitiation{
+			RefundTimestamp: big.NewInt(int64(contract.LockTime)),
+			SecretHash:      secretHash,
+			Participant:     common.HexToAddress(contract.Address),
+			Value:           big.NewInt(0).Mul(bigVal, BigGweiFactor),
+		})
+	}
+	return c.ETHSwap.Initiate(txOpts, inits)
+}
+
+func (c *contractorV0) Redeem(txOpts *bind.TransactOpts, redemptions []*asset.Redemption) (*types.Transaction, error) {
+	redemps := make([]ETHSwapRedemption, 0, len(redemptions))
+	for _, r := range redemptions {
+		secretB, secretHashB := r.Secret, r.Spends.SecretHash
+		if len(secretB) != 32 || len(secretHashB) != 32 {
+			return nil, fmt.Errorf("invalid secret and/or secret hash sizes, %d and %d", len(secretB), len(secretHashB))
+		}
+		var secret, secretHash [32]byte
+		copy(secret[:], secretB)
+		copy(secretHash[:], secretHashB)
+		redemps = append(redemps, ETHSwapRedemption{
+			Secret:     secret,
+			SecretHash: secretHash,
+		})
+	}
+	return c.ETHSwap.Redeem(txOpts, redemps)
+}
+
+func (c *contractorV0) Swap(ctx context.Context, secretHash [32]byte) (*SwapState, error) {
+	callOpts := &bind.CallOpts{
+		Pending: true,
+		From:    c.acctAddr,
+		Context: ctx,
+	}
+	state, err := c.ETHSwap.Swap(callOpts, secretHash)
+	if err != nil {
+		return nil, err
+	}
+
+	return &SwapState{
+		BlockHeight: state.InitBlockNumber.Uint64(),
+		LockTime:    time.Unix(state.RefundBlockTimestamp.Int64(), 0),
+		Secret:      state.Secret,
+		Initiator:   state.Initiator,
+		Participant: state.Participant,
+		Value:       state.Value.Uint64(),
+		State:       SwapStep(state.State),
+	}, nil
+}
+
+func (c *contractorV0) IsRedeemable(secretHash, secret [32]byte) (bool, error) {
+	return c.ETHSwap.IsRedeemable(&bind.CallOpts{From: c.acctAddr}, secretHash, secret)
+}
+
+func (c *contractorV0) EstimateRedeemGas(ctx context.Context, secrets [][32]byte) (uint64, error) {
+	redemps := make([]ETHSwapRedemption, 0, len(secrets))
+	for _, secret := range secrets {
+		redemps = append(redemps, ETHSwapRedemption{
+			Secret:     secret,
+			SecretHash: sha256.Sum256(secret[:]),
```

Comment:
These couldn't be fake? I think there was some discussion about this but I forget. EstimateInit is using fake secret hashes.

Reply:
Apparently not.

Reply:
It was wrong with random bytes?  I would expect only the length to be important, not the content.